src/item.cpp

```diff
@@ -2373,57 +2395,127 @@ std::string item::tname( unsigned int quantity, bool with_prefix ) const
-        amount = get_remaining_chapters( g->u );
-    } else if( ammo_capacity() > 0 ) {
-        // anything that can be reloaded including tools, magazines, guns and auxiliary gunmods
-        amount = ammo_remaining();
-        show_amt = true;
-    } else if( count_by_charges() && !has_infinite_charges() ) {
-        // A chargeable item
-        amount = charges;
-    }
-
-    if( amount || show_amt ) {
-        if( ammo_type() == "money" ) {
-            amt = string_format( " ($%.2f)", ( double ) amount / 100 );
+
+    if ( naming_style == NAMING_CONTAINER_LEADING || naming_style == NAMING_CONTAINER_TRAILING ) {
+        // This is recursive so keep an eye on it, could be a source of
+        // trouble if container type items continue to become more complex
+        if( !this->contents.empty() ) {
+            content_text = this->contents.front().display_name( quantity );
         } else {
-            amt = string_format( " (%i)", amount );
+            debugmsg("Tried to generate container name with contents for empty container.");
+            content_text = "????";
         }
     }
+    
+    // Generate charges text, ex: (0) (13) ($25.50)
+    if( naming_style == NAMING_WITH_CHARGES || naming_style == NAMING_WITH_AMMO_TYPE ) {
+        if ( this->is_book() ){
+            charges = this->get_remaining_chapters( g->u );
+        } else if ( this->ammo_capacity() > 0 ) {
+            charges = this->ammo_remaining();
+        } else {
+            charges = this->charges;
+        }
 
-    return string_format( "%s%s%s", name.c_str(), sidetxt.c_str(), amt.c_str() );
+        if( this->ammo_type() == "money" ) {
+            // Charges are money in cents; reformat to display as dollars
+            charges_text = string_format( " ($%.2f)", ( double ) charges / 100 );
+        } else {
+            // All other charges displayed as number with no units
+            charges_text = string_format( " (%i)", charges );
+        }
+    }
+    
+    // Generate ammo type text, ex: (9x19mm FMJ) (00 Shot)
+    if( naming_style == NAMING_WITH_AMMO_TYPE ) {
+        if ( this->ammo_data() ) {
+            // Most guns use ammo_data()
+            ammo_type_text = string_format( " (%s)", this->ammo_data()->nname( 1 ).c_str() );
+        } else if ( this->curammo ) {
+            // Some weapons use curammo instead, especially those that
+            // don't use magazines
+            ammo_type_text = string_format( " (%s)", this->curammo->nname( 1 ).c_str() );
+        } else if ( !this->contents.empty() && this->contents.front().is_ammo() ) {
+            // Any other item containing ammo that doesn't actually use
+            // the ammo and we still want the ammo type (bandoliers)
+            ammo_type_text = string_format( " (%s)", this->contents.front().label( 1 ) );
+        }
+    }
+    
+      
+    if ( naming_style == NAMING_STANDARD ) {
+        return string_format( _( "%1$s" ), item_text.c_str() );
```

Comment:
A general note: `string_format` and other functions that do string formatting understand `std::string` now, so you can use a `std::string` directly, without calling `.c_str()`. E.g. `string_format( "%s", item_text )` will work fine.